File: lesson02/SparkRDD/src/main/java/WordCount.java
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.broadcast.Broadcast;
import scala.Tuple2;

import java.util.Arrays;

public class WordCount {
    public static void main(String[] args) {
        final String input = args[0];
        final String output = args[1];

        final String delimiter = args.length > 2 ? args[3] : " ";

        JavaSparkContext sc = new JavaSparkContext();

        Broadcast<String> broadcasted = sc.broadcast(delimiter);

        sc.textFile(input)
                .flatMap(line -> Arrays.asList(line.split(broadcasted.getValue())).iterator())
                .mapToPair(word -> new Tuple2<>(word, 1))
                .reduceByKey(Integer::sum)
                .map(tuple -> tuple._1 + "\t" + tuple._2)
                .saveAsTextFile(output);

        sc.stop();
    }
}

File: lesson02/ScalaSparkRDD/src/main/scala/WordCount.scala
import org.apache.spark.SparkContext

object WordCount {
  def main(args: Array[String]): Unit = {
    val input = args(0)
    val output = args(1)
    val sc = new SparkContext()
    sc.textFile(input)
      .flatMap(line => line.split("\\s"))
      .map(word => (word.toString.toLowerCase.replaceAll("[\\\",\\.\\!\\?()\\:#\\/]|\\s{2,}", ""), 1))
      .reduceByKey(_ + _) // .reduceByKey((x, y) => x + y)
      .sortBy { case (word, count) => count } // .sortBy(_._2) or .sortBy(pair => pair._2)
      .map(tuple => tuple._2 + "; \"" + tuple._1 + "\"")
      .saveAsTextFile(output)
    sc.stop()
  }
}